### PR TITLE
List Library WIP

### PR DESCRIPTION
**Pull request by liampalmer43**
- PR for the List Library

### CODE REVIEW

File: main/src/library/List.flix [hunk 1]
```diff
@@ -0,0 +1,527 @@
...
+    /// If `n < 0` or `n > length(xs)` the semantics are undefined.
+    ///
+    def rotateLeft[a](n: Int32, xs: List[a]): List[a] = append(suffix(length(xs)-n, xs), prefix(n, xs))
+
+    ///
+    /// Returns `xs` with its elements rotated `n` positions to the right.
+    /// That is, returns a new list where the last `n` elements in `xs` are the first `n` elements of the new list.
+    /// If `n < 0` or `n > length(xs)` the semantics are undefined.
+    ///
+    def rotateRight[a](n: Int32, xs: List[a]): List[a] = append(suffix(n, xs), prefix(length(xs)-n, xs))
```
Comment (magnus-madsen) [inline main/src/library/List.flix]: As above.

File: main/src/library/List.flix [hunk 2]
```diff
@@ -0,0 +1,527 @@
...
+        case (0, r :: rs) => x :: rs
+        case (i, r :: rs) => r :: replace(i-1, x, rs)
+    }
+
+    ///
+    /// Returns `ys` with the `n` elements starting at position `p` replaced with elements from `xs`.
+    /// If `n = 0` then `ys` is returned.
+    /// Otherwise, if `n < 0` or `p < 0` or `p + n > length(xs)` the semantics are undefined.
+    ///
+    def patch[a](p: Int32, n: Int32, xs: List[a], ys: List[a]): List[a] = match (p, n, xs, ys) with {
```
Comment (magnus-madsen) [inline main/src/library/List.flix]: I want to think more about this function. Two questions for you:

(1) How do you "read" this function? For example, when reading `map(f, xs)` you say: "map (i.e. apply) the function `f` over the elements `xs`.
(2) Is it possible to make this function work when some of the arguments are "out-of-bounds" or does that not make sense?

File: main/src/library/List.flix [hunk 3]
```diff
@@ -0,0 +1,527 @@
...
+    def subsequences[a](xs: List[a]): List[List[a]] = match xs with {
+        case Nil => Nil :: Nil
+        case x :: rs => append(apply(x, subsequences(rs)), subsequences(rs))
+    }
+
+    ///
+    /// Helper function for `permutations` and `subsequences`.
+    /// Returns `xs` with `x` added to the beginning of each element of `xs`.
+    ///
+    def apply[a](x: a, ps: List[List[a]]): List[List[a]] = match ps with {
```
Comment (magnus-madsen) [inline main/src/library/List.flix]: Add `helper` to the name, and then we will later inline it.

File: main/src/library/List.flix [hunk 4]
```diff
@@ -0,0 +1,527 @@
...
+////    - If the list is empty the semantics are undefined.
+////- **max(xs: List[Int]): Int**
+////    - Returns the maximum element of a non-empty list.
+////    - If the list is empty the semantics are undefined.
+////- **minBy(f: (A, A) => Bool, xs: List[Int]): Int**
+////    - Returns the minimum element of a non-empty list according the total order defined by the predicate f. That is, if f(x, y) == true then x is smaller or equal to y. Two elements are equal if f(x, y) = f(y, x).
+////    - If the list is empty the semantics are undefined.
+////- **maxBy(f: (A, A) => Bool, xs: List[Int]): Int**
+////    - Similar to minBy but returns the largest element.
+////- **sort(xs: List[Int]): Int**
```
Comment (magnus-madsen) [inline main/src/library/List.flix]: No sorting. Don't try to implement this one.

File: main/src/library/List.flix [hunk 5]
```diff
@@ -0,0 +1,527 @@
...
+////    - Returns the longest prefix of xs that satisfy the predicate f.
+////
+////## Aggregation and Sorting
+////- **sum(xs: List[Int]): Int**
+////    - Returns the sum of the elements in the list.
+////    - Returns zero if the list is empty.
+////- **product(xs: List[Int]): Int**
+////    - Returns the product of the elements in the list.
+////    - Returns one if the list is empty.
+////- **min(xs: List[Int]): Int**
```
Comment (magnus-madsen) [inline main/src/library/List.flix]: Skip for now

File: main/src/library/List.flix [hunk 6]
```diff
@@ -0,0 +1,527 @@
...
+////    - Returns the first n elements of xs.
+////    - Returns the entire list if n is larger than the length of the list.
+////- **takeWhile(f: A => Bool, xs: List[A]): List[A]**
+////    - Returns the longest prefix of xs that satisfy the predicate f.
+////
+////## Aggregation and Sorting
+////- **sum(xs: List[Int]): Int**
+////    - Returns the sum of the elements in the list.
+////    - Returns zero if the list is empty.
+////- **product(xs: List[Int]): Int**
```
Comment (magnus-madsen) [inline main/src/library/List.flix]: Skip for now

File: main/src/library/List.flix [hunk 7]
```diff
@@ -0,0 +1,527 @@
...
+////- **dropWhile(f: A => Bool, xs: List[A]): List[A]**
+////    - Returns the list except for the longest prefix that satisfy the predicate f.
+////- **take(n: Int, xs: List[A]): List[A]**
+////    - Returns the first n elements of xs.
+////    - Returns the entire list if n is larger than the length of the list.
+////- **takeWhile(f: A => Bool, xs: List[A]): List[A]**
+////    - Returns the longest prefix of xs that satisfy the predicate f.
+////
+////## Aggregation and Sorting
+////- **sum(xs: List[Int]): Int**
```
Comment (magnus-madsen) [inline main/src/library/List.flix]: Skip for now

File: main/src/library/List.flix [hunk 8]
```diff
@@ -0,0 +1,527 @@
...
+////    - Returns the sum of the elements in the list.
+////    - Returns zero if the list is empty.
+////- **product(xs: List[Int]): Int**
+////    - Returns the product of the elements in the list.
+////    - Returns one if the list is empty.
+////- **min(xs: List[Int]): Int**
+////    - Returns the minimum element of a non-empty list.
+////    - If the list is empty the semantics are undefined.
+////- **max(xs: List[Int]): Int**
```
Comment (magnus-madsen) [inline main/src/library/List.flix]: Skip for now

File: main/src/library/List.flix [hunk 9]
```diff
@@ -0,0 +1,527 @@
...
+////    - Returns the maximum element of a non-empty list.
+////    - If the list is empty the semantics are undefined.
+////- **minBy(f: (A, A) => Bool, xs: List[Int]): Int**
```
Comment (magnus-madsen) [inline main/src/library/List.flix]: Skip for now

File: main/src/library/List.flix [hunk 10]
```diff
@@ -0,0 +1,527 @@
...
+////    - Returns the minimum element of a non-empty list according the total order defined by the predicate f. That is, if f(x, y) == true then x is smaller or equal to y. Two elements are equal if f(x, y) = f(y, x).
+////    - If the list is empty the semantics are undefined.
+////- **maxBy(f: (A, A) => Bool, xs: List[Int]): Int**
```
Comment (magnus-madsen) [inline main/src/library/List.flix]: Skip for now

File: main/src/library/List.flix [hunk 11]
```diff
@@ -0,0 +1,527 @@
...
+////    - Similar to minBy but returns the largest element.
+////- **sort(xs: List[Int]): Int**
+////    - Returns the elements of the list sorted according to the natural order.
+////- **sortBy(f: (A, A) => Bool, xs: List[Int]): Int**
```
Comment (magnus-madsen) [inline main/src/library/List.flix]: Skip for now

File: main/src/library/List.flix [hunk 12]
```diff
@@ -0,0 +1,527 @@
...
+////    - Returns the elements of the list sorted according to the total order defined by the predicate f. That is, if f(x, y) == true then x is smaller or equal to y. Two elements are equal if f(x, y) = f(y, x).
+////- **groupBy(f: (A, A) => Bool, xs: List[A]): List[List[A]]**
```
Comment (magnus-madsen) [inline main/src/library/List.flix]: This one would be nice.

File: main/src/library/List.flix [hunk 13]
```diff
@@ -0,0 +1,527 @@
...
+////    - Pairwise applies the function f to the elements of the lists xs and ys and concatenates the result in one list. Functionally equivalent to flatten(map2(f, xs, ys)).
+////- **fold2(f: (C, A, B) => C, c: C, xs: List[A], ys: List[B]): C**
+////    - Alias for foldLeft2.
+////- **foldLeft2(f: (C, A, B) => C, c: C, xs: List[A], ys: List[B]): C**
+////    - Accumulates the result of applying the function f pairwise to the elements of xs and ys starting with the initial value c going from left to right. Functionally similar to foldLeft(f, c, zip(xs, ys)).
+////- **foldRight2(f: (A, B, C) => C, c: C, xs: List[A], ys: List[B]): C**
+////    - Similar to foldLeft2 but goes from the right to the left.
+////
+////## Combined Operations
+////- **concatMap(f: A => List[B], xs: List[A]): List[B]**
```
Comment (magnus-madsen) [inline main/src/library/List.flix]: skip

File: main/src/library/List.flix [hunk 14]
```diff
@@ -0,0 +1,527 @@
...
+////- **findMap(f: A => Opt[B], xs: List[A]): Opt[B]**
+////    - Collects the first result of applying the partial function f to each element of the list.
+////
+////## List Conversions
+////- **toMap(xs: List[(A, B)]): Map[A, B]**
+////    - Returns the association list as a map. If the list contains multiple values for the same key, the value appearing latest in the list is used.
+////- **toSet(xs: List[A]): Set[A]**
+////    - Returns the list as a set.
+////
+////## Order and Lattice Operations
```
Comment (magnus-madsen) [inline main/src/library/List.flix]: You can skip all of these lattices and order ops.

**Comment by magnus-madsen** [comment]
A general questions:

We have special syntax for `Cons` namely `::`. Would it be helpful to have some syntax for `append`? The two common choices are usually `:::` (scala) or `@@` in other languages.

For example:

```
def f: List[Int] = (1 :: 2 :: Nil) @@ (3 :: 4 :: 5 :: Nil)
```

File: main/src/library/List.flix [hunk 15]
```diff
@@ -0,0 +1,527 @@
...
+    /// Returns `ys` appended to `xs`.
+    ///
+    def append[a](xs: List[a], ys: List[a]): List[a] = match xs with {
+        case Nil => ys
+        case x :: rs => x :: append(rs, ys)
+    }
+
+    ///
+    /// Returns the element at position `p` in `xs`.
+    /// If `p` < 0 or `p` > length(xs)-1 the semantics are undefined.
```
Comment (mhyee) [inline main/src/library/List.flix]: @magnus-madsen Do we want functions with undefined semantics? Or are we going to have some sort of error?

File: main/src/library/List.flix [hunk 16]
```diff
@@ -0,0 +1,527 @@
...
+    ///
+    def at[a](p: Int32, xs: List[a]): a = match (p, xs) with {
```
Comment (mhyee) [inline main/src/library/List.flix]: There doesn't seem to be a good naming convention for this. Scala uses `apply` (i.e. parenthesis indexing), Haskell uses `!!`, OCaml uses `nth`, Racket doesn't seem to have it. C++ has `at` for vectors.

File: main/src/library/List.flix [hunk 17]
```diff
@@ -0,0 +1,527 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace List {
+
+    ///
+    /// Returns true iff `xs` is the empty list, i.e. `Nil`.
+    ///
+    def null[a](xs: List[a]): Bool = match xs with {
```
Comment (mhyee) [inline main/src/library/List.flix]: Another note about naming convention: Haskell typically uses all lowercase names, while Scala will use camelCase. What do we want in Flix?

Comment (magnus-madsen) [inline main/src/library/List.flix]: CamelCase

File: main/src/library/List.flix [hunk 18]
```diff
@@ -0,0 +1,579 @@
...
+    /// Returns `ys` appended to `xs`.
+    ///
+    def append[a](xs: List[a], ys: List[a]): List[a] = match xs with {
+        case Nil => ys
+        case x :: rs => x :: append(rs, ys)
+    }
+
+    ///
+    /// Returns the element at index `i` in `xs`.
+    /// If `i < 0` or `i > length(xs)-1` the semantics are undefined.
```
Comment (liampalmer43) [inline main/src/library/List.flix]: I'm unsure what should be done when the index is out of range here.

File: main/src/library/List.flix [hunk 19]
```diff
@@ -0,0 +1,579 @@
...
+    def at[a](i: Int32, xs: List[a]): a = match (i, xs) with {
+        case (0, x :: _) => x
+        case (p, x :: rs) => at(p - 1, rs)
+    }
+
+    ///
+    /// Returns `xs` with the first occurrence of `x` removed.
+    /// Returns `xs` if `x` does not exist in `xs`.
+    ///
+    def remove[a](x: a, xs: List[a]): List[a] = match xs with {
```
Comment (liampalmer43) [inline main/src/library/List.flix]: `remove` now removes the first occurrence of an element if there is one.  There is also `removeIndex` (below) which removes an element by index.  However, I can make this a helper function for `permutations` if it should not be a library function.

File: main/src/library/List.flix [hunk 20]
```diff
@@ -0,0 +1,527 @@
...
+    }
+
+    ///
+    /// Accumulates the result of applying `f` to `xs` going right to left.
+    /// That is, the result is of the form: `... f(xn-1, f(xn, s)) :: f(xn, s) :: s`.
+    ///
+    def scanRight[a,b](f: (a, b) -> b, s: b, xs: List[a]): List[b] = match xs with {
+        case Nil => s :: Nil
+        case x :: rs => match scanRight(f, s, rs) with {
+            case y :: ys => f(x, y) :: y :: ys
```
Comment (liampalmer43) [inline main/src/library/List.flix]: `scanRight` will never return Nil, so I didn't match it here.

File: main/src/library/List.flix [hunk 21]
```diff
@@ -0,0 +1,579 @@
...
+        case Nil => acc
+        case x :: rs => reverseHelper(rs, x :: acc)
+    }
+
+    ///
+    /// Returns `xs` with its elements rotated `n` positions to the left.
+    /// That is, returns a new list where the first `n mod length(xs)` elements in `xs`
+    /// are the last `n mod length(xs)` elements of the new list.
+    ///
+    def rotateLeft[a](n: Int32, xs: List[a]): List[a] =
```
Comment (liampalmer43) [inline main/src/library/List.flix]: `rotateLeft` and `rotateRight` are now valid for any `n`.  The distance rotated is the positive value of `n mod length(xs)`.

File: main/src/library/List.flix [hunk 22]
```diff
@@ -0,0 +1,579 @@
...
+        case (p, r :: rs) => r :: replace(p-1, x, rs)
+    }
+
+    ///
+    /// Returns `ys` with the `n` elements starting at index `i` replaced with the elements of `xs`.
+    /// If any of the indices `i, i+1, i+2, ... , i+n-1` are out of range in `ys` then no patching is done at these indices.
+    /// If `xs` becomes depleted then no further patching is done.
+    /// If patching occurs at index `i+j` in `ys`, then the element at index `j` in `xs` is used.
+    ///
+    def patch[a](i: Int32, n: Int32, xs: List[a], ys: List[a]): List[a] = patchHelper(i, n, drop(-i, xs), ys, 0)
```
Comment (liampalmer43) [inline main/src/library/List.flix]: `patch` is now valid for any set of inputs.  All or part of `xs` is simply mapped/patched into a range of `ys`.  Anywhere the range is invalid, no patching occurs.

File: main/src/library/List.flix [hunk 23]
```diff
@@ -0,0 +1,579 @@
...
+    ///
+    def mapWithIndexHelper[a,b](f: (a, Int32) -> b, xs: List[a], i: Int32): List[b] = match xs with {
+        case Nil => Nil
+        case x :: rs => f(x, i) :: mapWithIndexHelper(f, rs, i+1)
+    }
+
+    ///
+    /// Returns the result of applying `f` to every element in `xs` and concatenating the results.
+    ///
+    def flatMap[a,b](f: a -> List[b], xs: List[a]): List[b] = match xs with {
```
Comment (liampalmer43) [inline main/src/library/List.flix]: I believe this is O(n^2) if `xs` is length `n` and `f` outputs a list of length `n`, assuming that appending a list of length `n` onto another list takes O(n) time.  My analysis might be wrong though, I'm unsure how long appending takes in Flix.

File: main/src/library/List.flix [hunk 24]
```diff
@@ -0,0 +1,527 @@
...
+    ///
+    /// Alias for `scanLeft`.
+    ///
+    def scan[a,b](f: (b, a) -> b, s: b, xs: List[a]): List[b] = scanLeft(f, s, xs)
+
+    ///
+    /// Accumulates the result of applying `f` to `xs` going left to right.
+    /// That is, the result is of the form: `s :: f(s, x1) :: f(f(s, x1), x2)  ...`.
+    ///
+    def scanLeft[a,b](f: (b, a) -> b, s: b, xs: List[a]): List[b] = s :: scanLeftHelper(f, s, xs)
```
Comment (liampalmer43) [inline main/src/library/List.flix]: Sounds good, I'll inline the helper functions when this is implemented.

File: main/src/library/List.flix [hunk 25]
```diff
@@ -0,0 +1,612 @@
...
+        case Nil => Nil
+        case y :: Nil => y
+        case y1 :: y2 :: rs => append(append(y1, xs), intercalate(xs, y2 :: rs))
+    }
+
+    ///
+    /// Returns the transpose of `xs`.
+    /// Returns `xs` if the dimensions of the elements of `xs` are mismatched.
+    ///
+    def transpose[a](xs: List[List[a]]): List[List[a]] = if (!uniformHelper(xs)) xs else match xs with {
```
Comment (liampalmer43) [inline main/src/library/List.flix]: As of now, in the case that the dimensions of `xs` are mismatched, `xs` is returned.  Should some sort of error occur instead?

File: main/src/library/List.flix (hunk 20 above)
Comment (magnus-madsen) [inline main/src/library/List.flix]: Oh, I see

File: main/src/library/List.flix (hunk 18 above)
Comment (magnus-madsen) [inline main/src/library/List.flix]: Yeah, in this case I don't think anything can be done.

File: main/src/library/List.flix (hunk 19 above)
Comment (magnus-madsen) [inline main/src/library/List.flix]: If you could add the `@internal` annotation on all helper functions, then I will hide them in the flixdoc.

File: main/src/library/List.flix [hunk 26]
```diff
@@ -0,0 +1,579 @@
...
+        case x :: rs => if (f(x)) Some(x) else findLeft(f, rs)
+    }
+
+    ///
+    /// Optionally returns the first element of `xs` that satisfies the predicate `f` when searching from right to left.
+    ///
+    def findRight[a](f: a -> Bool, xs: List[a]): Option[a] = match xs with {
+        case Nil => None
+        // Unable to add this now, I'm unsure how to access the Option namespace.
+        // case x :: rs => Option/withDefault(findRight(f, rs), if (f(x)) Some(x) else None)
```
Comment (magnus-madsen) [inline main/src/library/List.flix]: Did you include `Option.flix` in the test case/test runner?

File: main/src/library/List.flix (hunk 23 above)
Comment (magnus-madsen) [inline main/src/library/List.flix]: Yes, I believe you are correct!

File: main/src/library/List.flix [hunk 27]
```diff
@@ -0,0 +1,748 @@
...
+                            case None => filterMap(f, rs)
+                            case Some(v) => v :: filterMap(f, rs)
+        }
+    }
+
+    ///
+    /// Returns the first non-None result of applying the partial function `f` to each element of `xs`.
+    /// Returns `None` if every element of `xs` is `None`.
+    ///
+    def findMap[a,b](f: a -> Option[b], xs: List[a]): Option[b] = match xs with {
```
Comment (liampalmer43) [inline main/src/library/List.flix]: Is this the correct functionality?

**Comment by magnus-madsen** [comment]
Hello Liam. This looks super-good. I am going to do one last review (expecting minor changes) and then hopefully we can merge it in. Just give me a few days :)

File: main/src/library/List.flix (hunk 27 above)
Comment (magnus-madsen) [inline main/src/library/List.flix]: This looks correct. If you have two lists: 1, 2, 3 and 5, 6, I would expect that when folding right you align the lists like so:

```
1 2 3
   5 6
```

and so you would apply the function to (3, 6) and then to (2, 5) and then  you would be done.

File: main/src/library/List.flix [hunk 28]
```diff
@@ -0,0 +1,756 @@
...
+    }
+
+    ///
+    /// Returns the concatenation of the elements in `ys` with the elements of `xs` inserted between every two adjacent elements.
+    /// That is, returns y1 + x + y2 + x + ... + x + yn.
+    ///
+    def intercalate[a](xs: List[a], ys: List[List[a]]): List[a] = match ys with {
+        case Nil => Nil
+        case y :: Nil => y
+        case y1 :: y2 :: rs => append(append(y1, xs), intercalate(xs, y2 :: rs))
```
Comment (magnus-madsen) [inline main/src/library/List.flix]: You can use `:::` for append to make the code easier to read.

Comment (magnus-madsen) [inline main/src/library/List.flix]: E.g. ```myList ::: someOtherList```

File: main/src/library/List.flix [hunk 29]
```diff
@@ -0,0 +1,756 @@
...
+    //
+    // ## Special Folds
+    //
+
+    ///
+    /// Returns the number of elements in `xs` that satisfy the predicate `f`.
+    ///
+    def count[a](f: a -> Bool, xs: List[a]): Int32 = match xs with {
+        case Nil => 0
+        case x :: rs => let r = count(f, rs);
```
Comment (magnus-madsen) [inline main/src/library/List.flix]: I think our preferred style is to insert an extra line break, like so:

```
case x :: rs => 
  let r = count(f, rs);
    if (f(x)) r+1 else r
```

File: main/src/library/List.flix [hunk 30]
```diff
@@ -0,0 +1,756 @@
...
+                            if (f(x)) r+1 else r
+    }
+
+    ///
+    /// Returns the concatenation of the elements in `xs`.
+    ///
+    def concat[a](xs: List[List[a]]): List[a] = match xs with {
+        case Nil => Nil
+        case x :: rs => append(x, concat(rs))
```
Comment (magnus-madsen) [inline main/src/library/List.flix]: Use `:::`

File: main/src/library/List.flix [hunk 31]
```diff
@@ -0,0 +1,756 @@
...
+    //
+    // ## Sub Lists
+    //
+
+    ///
+    /// Returns a list of every element in `xs` that satisfies the predicate `f`.
+    ///
+    def filter[a](f: a -> Bool, xs: List[a]): List[a] = match xs with {
+        case Nil => Nil
+        case x :: rs => let r = filter(f, rs);
```
Comment (magnus-madsen) [inline main/src/library/List.flix]: line break before let.

File: main/src/library/List.flix (hunk 27 above)
Comment (magnus-madsen) [inline main/src/library/List.flix]: `findMap` should combine find and map. So you give it a function `a -> Option[b]` when this function returns `Some(...)` we will have found the thing we are interested in. So yeah, its look correct to me.

**Comment by magnus-madsen** [comment]
Great job! 

I am merging this into master. I added some minor comments which can be addressed in a future pull request. Once the Set and Map implementations are complete I plan to do another review of the entire library to look for improvements. See #255 .

**Pull request merged**